src/main/java/org/spongepowered/common/data/builder/tileentity/AbstractTileEntityBuilder.java

```diff
@@ -0,0 +1,94 @@
+import net.minecraft.block.Block;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.block.BlockType;
+import org.spongepowered.api.block.entity.BlockEntity;
+import org.spongepowered.api.block.entity.BlockEntityType;
+import org.spongepowered.api.data.persistence.*;
+import org.spongepowered.api.world.server.ServerWorld;
+import org.spongepowered.common.accessor.tileentity.TileEntityTypeAccessor;
+import org.spongepowered.common.util.Constants;
+
+import java.util.Optional;
+import java.util.Set;
+
+public abstract class AbstractTileEntityBuilder<T extends BlockEntity> extends AbstractDataBuilder<T> implements DataBuilder<T> {
+    protected AbstractTileEntityBuilder(Class<T> requiredClass, int supportedVersion) {
+        super(requiredClass, supportedVersion);
+    }
+
+    @Override
+    protected Optional<T> buildContent(DataView container) throws InvalidDataException {
+        Preconditions.checkNotNull(container);
+
+        if (!container.contains(Constants.TileEntity.TILE_TYPE, Constants.TileEntity.WORLD, Queries.POSITION_X, Queries.POSITION_Y, Queries.POSITION_Z)) {
+            return Optional.empty();
+        }
+
+        String worldName = container.getString(Constants.TileEntity.WORLD).get();
+        Optional<ServerWorld> worldOptional = Sponge.getGame().getServer().getWorldManager().getWorld(worldName);
+        if (!worldOptional.isPresent()) {
+            throw new InvalidDataException("The provided container references a world that does not exist!");
+        }
+
+        final Optional<BlockEntityType> type = container.getCatalogType(Constants.TileEntity.TILE_TYPE, BlockEntityType.class);
+        if (!type.isPresent()) {
+            // TODO: Double check, do we want to throw an InvalidDataException since the catalog type is not registered at all?
+            return Optional.empty();
+        }
+
+        TileEntityTypeAccessor accessor = (TileEntityTypeAccessor) type.get();
+        Set<Block> validBlocks = accessor.accessor$getValidBlocks();
+
+        if (validBlocks.isEmpty()) {
+            return Optional.empty(); // This should not be possible, but the TileEntityType.Builder does not strictly enforce it.
+        }
+
+        // TODO: Do we want to always use the first entry for this? Hardcoding this is an option but it would result in a similar issue with stable-7 where we still need a way to handle this with modded additions.
+        BlockType blockType = (BlockType) validBlocks.iterator().next();
+
+        // Now we should be ready to actually translate the TileEntity with the right block.
+
+        final int x = container.getInt(Constants.DataSerializers.X_POS).get();
+        final int y = container.getInt(Constants.DataSerializers.Y_POS).get();
+        final int z = container.getInt(Constants.DataSerializers.Z_POS).get();
+
+        worldOptional.get().getLocation(x, y, z).setBlockType(blockType);
```

Comment:
I'm gonna put a stop right here and say "no." We kinda have to work with deserialization failing when the setBlockType is cancelled by other plugins. If it's cancelled, then it should return empty fast.